### ISSUE TITLE
test loss and visualizing estimated values

### ISSUE THREAD

**Issue by ttsesm** [nikogamulin/torch-points3d#1]
Hi @nikogamulin, 

I wanted to ask you what should be the expected testing loss for the corresponding dataset. I've noticed [in your graphs in the issue that you have opened in the torch-points3d](https://github.com/nicolas-chaulet/torch-points3d/issues/502#issue-748191795) repository that you also get these flat graph (with the spikes of course). However, shouldn't be the test loss also following the decreasing pattern of the train loss?

In my dataset I'm also getting something similar and I am trying to figure out if this is expected or not, any idea? Actually if I zoom in I have a decreasing behavior though it is kind of small.

Here are some preliminary runs of my regression script (I am using the sparseconv3d network by the way and not the Rsconv as in your case):

![losses](https://user-images.githubusercontent.com/10189018/100801893-8468f180-3420-11eb-95c9-f80705ee14b3.png)

Also another question is how did you access the estimated values in order to plot the graphs that you have shown [here](https://github.com/nicolas-chaulet/torch-points3d/issues/502#issuecomment-733014030). I do not know if you could possibly give an example as well.

Thanks.

**Comment by nikogamulin** [comment]
hi @ttsesm , after modifying the options it seems like the spikes are gone. I modified the superquadrics.py script - please try using the same options on your dataset and let me know the results.

For graphs, please check superquadrics_regression_tracker.py sample

**Comment by ttsesm** [comment]
Good to hear, I saw also the results that you posted in the other thread.

In my case your options it doesnt seem to be useful. I do not know what is going on, I mean with the `RSConv` model the output is totally wrong. On the other hand with the `Sparseconv3d` model what I've noticed is that all the regressed values are more or less the same for all my points and I am trying to figure out why is this happening.

This my code so far:
```
import os
import torch
import pickle
import torch.nn as nn
import csv
from glob import glob
import numpy as np
import pandas as pd
from tqdm import tqdm

import datetime
import time

from torch_points3d.models.base_model import BaseModel
from torch_points3d.datasets.base_dataset import BaseDataset
from torch_points3d.applications.sparseconv3d import SparseConv3d
from torch_points3d.applications.rsconv import RSConv
# from torch_points3d.metrics.regression_tracker import RegressionTracker
from torch_points3d.core.data_transform import GridSampling3D

from torch.utils.data import Dataset
from torch_geometric.data import Batch, Data, InMemoryDataset, DataLoader

from torch_points3d.metrics.colored_tqdm import Coloredtqdm as Ctq

from omegaconf import OmegaConf

from torch.utils.tensorboard import SummaryWriter

from typing import Dict
import torchnet as tnt

from torch_points3d.metrics.confusion_matrix import ConfusionMatrix
from torch_points3d.metrics.base_tracker import BaseTracker, meter_value
from torch_points3d.models import model_interface

def read_pcd_data(filename):
    """read XYZ, lux and light intensity for each vertex."""
    assert os.path.isfile(filename)

    data = pd.read_csv(filename, header=None, delimiter=',', low_memory=False).to_numpy(dtype='str')

    return data

def export(pcd_file):
    pcd_data = read_pcd_data(pcd_file)

    return pcd_data.astype(np.float32)

class RegressionTracker(BaseTracker):
    def __init__(self, dataset, stage="train", wandb_log=False, use_tensorboard: bool = False):
        """ This is a generic tracker for segmentation tasks.
        It uses a confusion matrix in the back-end to track results.
        Use the tracker to track an epoch.
        You can use the reset function before you start a new epoch
        Arguments:
            dataset  -- dataset to track (used for the number of classes)
        Keyword Arguments:
            stage {str} -- current stage. (train, validation, test, etc...) (default: {"train"})
            wandb_log {str} --  Log using weight and biases
        """
        super(RegressionTracker, self).__init__(stage, wandb_log, use_tensorboard)
        self.reset(stage)

    def reset(self, stage="train"):
        super().reset(stage=stage)
        # self._loss_dimension = tnt.meter.AverageValueMeter()
        # self._loss_epsilon = tnt.meter.AverageValueMeter()
        # self._loss_offset = tnt.meter.AverageValueMeter()
        self._loss_lux = tnt.meter.AverageValueMeter()

    @staticmethod
    def detach_tensor(tensor):
        if torch.torch.is_tensor(tensor):
            tensor = tensor.detach()
        return tensor

    @staticmethod
    def compute_loss_by_components(y_hat, y):
        # # y = torch.reshape(y, (12, 8))
        # dimensions = y[:, 0:3]
        # epsilons = y[:, 3:5]
        # offsets = y[:, 5:]
        #
        # dimensions_hat = y_hat[:, 0:3]
        # epsilons_hat = y_hat[:, 3:5]
        # offsets_hat = y_hat[:, 5:]
        #
        # diff_dimensions = dimensions - dimensions_hat
        # avg_loss_dimensions = torch.sum(diff_dimensions*diff_dimensions)/diff_dimensions.numel()
        #
        # diff_epsilons = epsilons - epsilons_hat
        # avg_loss_epsilons = torch.sum(diff_epsilons*diff_epsilons)/diff_epsilons.numel()
        #
        # diff_offsets = offsets - offsets_hat
        # avg_loss_offsets = torch.sum(diff_offsets*diff_offsets)/diff_offsets.numel()
        #
        # return avg_loss_dimensions, avg_loss_epsilons, avg_loss_offsets

        diff = y - y_hat
        avg_loss = torch.sum(diff*diff)/diff.numel()

        return avg_loss


    def track(self, model: model_interface.TrackerInterface, **kwargs):
        """ Add current model predictions (usually the result of a batch) to the tracking
        """
        super().track(model)

        outputs = model.get_output()
        # targets = model.get_labels().flatten()
        targets = model.get_labels()

        # avg_loss_dimensions, avg_loss_epsilons, avg_loss_offsets = self.compute_loss_by_components(outputs, targets)
        avg_loss = self.compute_loss_by_components(outputs, targets)

        # self._loss_dimension.add(avg_loss_dimensions.detach().cpu().numpy())
        # self._loss_epsilon.add(avg_loss_epsilons.detach().cpu().numpy())
        # self._loss_offset.add(avg_loss_offsets.detach().cpu().numpy())
        self._loss_lux.add(avg_loss.detach().cpu().numpy())

    def get_metrics(self, verbose=False) -> Dict[str, float]:
        """ Returns a dictionnary of all metrics and losses being tracked
        """
        metrics = super().get_metrics(verbose)
        # metrics["{}_loss_dimension".format(self._stage)] = meter_value(self._loss_dimension)
        # metrics["{}__loss_epsilon".format(self._stage)] = meter_value(self._loss_epsilon)
        # metrics["{}_loss_offset".format(self._stage)] = meter_value(self._loss_offset)
        metrics["{}_loss_lux".format(self._stage)] = meter_value(self._loss_lux)
        return metrics

    @property
    def metric_func(self):
        self._metric_func = {
            "acc": max,
        }  # Those map subsentences to their optimization functions
        return self._metric_func

class DeepLuxLoader(Dataset):
    # def __init__(self, dataset_size, points_count, dimension_max, do_normalize=True, transform=None):
    def __init__(self, files, type="Training files", transform=None):
        """
        Args:

        """
        super().__init__()

        self.pos = []
        self.x = []
        self.y = []
        self.coords = []
        dataset_size = len(files)
        pbar = tqdm(total=dataset_size, desc=type)

        idx = 0
        while idx < dataset_size:
            pcd = export(files[idx])
            try:
                points = pcd[:, 1:4]
                # _illum = pcd[:, 0]
                _illum = np.column_stack([pcd[:, 0], pcd[:, 5]])
                lux = pcd[:, 4]

            except:
                continue

            # self.pos.append(points)
            self.pos.append(points)
            self.x.append(_illum)
            self.y.append(lux)
            # self.coords.append(points*1000)
            self.coords.append(points)
            idx += 1
            pbar.update(1)
        pbar.close()

    def __len__(self):
        return len(self.pos)

    def __getitem__(self, idx):
        if torch.is_tensor(idx):
            idx = idx.tolist()

        x = torch.from_numpy(self.x[idx]).float()
        pos = torch.from_numpy(self.pos[idx]).float()
        y = torch.from_numpy(self.y[idx]).float()
        coords = torch.from_numpy(self.coords[idx]).int()
        # data = Data(pos=pos, x=x.unsqueeze(-1), y=y, coords=coords)
        data = Data(pos=pos, x=x, y=y, coords=coords)
        # data = Data(pos=pos, x=x, y=y)

        # tr = GridSampling3D(0.01, quantize_coords=True)
        # data = tr(data)

        return data

class DeepLuxDataset(BaseDataset):
    def __init__(self, dataset_opt):
        super().__init__(dataset_opt)

        csv_files = sorted(glob(self.dataset_opt.dataroot + '**/*.csv', recursive=True))
        # csv_files = csv_files[0:500]

        # Split files to training and testing.
        # Common split percentages include:
        # Train: 80 %, Test: 20 %
        # Train: 67 %, Test: 33 %
        # Train: 50 %, Test: 50 %
        dataset_sizes = np.array([self.dataset_opt.training.dataset_size, self.dataset_opt.testing.dataset_size])
        train_files, test_files = self.splitPerc(csv_files, dataset_sizes)

        self.train_dataset = DeepLuxLoader(train_files.tolist(), type="Training files", transform=self.train_transform)
        self.test_dataset = DeepLuxLoader(test_files.tolist(), type="Testing files", transform=self.test_transform)

    def get_tracker(self, wandb_log: bool, tensorboard_log: bool):
        """Factory method for the tracker
        Arguments:
            wandb_log - Log using weight and biases
            tensorboard_log - Log using tensorboard
        Returns:
            [BaseTracker] -- tracker
        """
        return RegressionTracker(self, wandb_log=wandb_log, use_tensorboard=tensorboard_log)

    def splitPerc(self, l, perc):
        # Turn percentages into values between 0 and 1
        splits = np.cumsum(perc) / 100.

        if splits[-1] != 1:
            raise ValueError("percents don't add up to 100")

        # Split doesn't need last percent, it will just take what is left
        splits = splits[:-1]

        # Turn values into indices
        splits *= len(l)

        # Turn double indices into integers.
        # CAUTION: numpy rounds to closest EVEN number when a number is halfway
        # between two integers. So 0.5 will become 0 and 1.5 will become 2!
        # If you want to round up in all those cases, do
        # splits += 0.5 instead of round() before casting to int
        splits = splits.round().astype(np.int)

        return np.split(l, splits)

class Sparseconv3dRegressor(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.backbone = SparseConv3d("unet", input_nc=2, output_nc=1, num_layers=4, backend="torchsparse")
        # self.head = torch.nn.Linear(self.backbone.output_nc, 1)  # output layer
        self.loss_function = torch.nn.L1Loss()
        # self.loss_function = torch.nn.MSELoss()

    @property
    def conv_type(self):
        """ This is needed by the dataset to infer which batch collate should be used"""
        return self.backbone.conv_type

    def get_output(self):
        """ This is needed by the tracker to get access to the ouputs of the network"""
        return self.output

    def get_labels(self):
        """ Needed by the tracker in order to access ground truth labels"""
        return self.labels

    def get_current_losses(self):
        """ Entry point for the tracker to grab the loss """
        return {"loss": float(self.loss)}

    def forward(self, data):
        # Set labels for the tracker
        self.labels = data.y.squeeze()

        # Forward through the network
        data_out = self.backbone(data)
        self.output = data_out.x.squeeze()

        # Set loss for the backward pass
        self.loss = self.loss_function(self.labels, self.output)

    def backward(self):
        self.loss.backward()

class RSConvRegressor(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.backbone = RSConv("unet", input_nc=2, output_nc=1, num_layers=4)
        # self.head = torch.nn.Linear(self.backbone.output_nc, 1)  # output layer
        self.loss_function = torch.nn.L1Loss()
        # self.loss_function = torch.nn.MSELoss()

    @property
    def conv_type(self):
        """ This is needed by the dataset to infer which batch collate should be used"""
        return self.backbone.conv_type

    def get_output(self):
        """ This is needed by the tracker to get access to the ouputs of the network"""
        return self.output

    def get_labels(self):
        """ Needed by the tracker in order to access ground truth labels"""
        return self.labels

    def get_current_losses(self):
        """ Entry point for the tracker to grab the loss """
        return {"loss": float(self.loss)}

    def forward(self, data):
        # Set labels for the tracker
        self.labels = data.y.squeeze()

        # Forward through the network
        data_out = self.backbone(data)
        self.output = data_out.x.squeeze()

        # Set loss for the backward pass
        self.loss = self.loss_function(self.labels, self.output)

    def backward(self):
        self.loss.backward()

def train_epoch(device):
    model.to(device)
    model.train()
    # tracker.reset("train")
    train_loader = dataset.train_dataloader
    # iter_data_time = time.time()
    losses = []
    with Ctq(train_loader) as tq_train_loader:
        for i, data in enumerate(tq_train_loader):
            # t_data = time.time() - iter_data_time
            # iter_start_time = time.time()
            optimizer.zero_grad()
            data.to(device)
            model.forward(data)
            losses.append(model.get_current_losses()["loss"])

            # loss = model.get_current_losses()["loss"]
            # print(loss)
            model.backward()
            optimizer.step()
            # scheduler.step()

            # if i % 10 == 0:
            #     tracker.track(model)
            #
            # tq_train_loader.set_postfix(
            #     **tracker.get_metrics(),
            #     data_loading=float(t_data),
            #     iteration=float(time.time() - iter_start_time),
            # )
            # iter_data_time = time.time()

    return sum(losses) / len(losses)

def test_epoch(device):
    model.to(device)
    model.eval()
    losses = []
    with torch.no_grad():
        # tracker.reset("test")
        test_loader = dataset.test_dataloaders[0]
        # iter_data_time = time.time()


        with Ctq(test_loader) as tq_test_loader:
            for i, data in enumerate(tq_test_loader):
                # t_data = time.time() - iter_data_time
                # iter_start_time = time.time()
                data.to(device)
                model.forward(data)
                losses.append(model.get_current_losses()["loss"])

                # tracker.track(model)

                # tq_test_loader.set_postfix(
                #     **tracker.get_metrics(),
                #     data_loading=float(t_data),
                #     iteration=float(time.time() - iter_start_time),
                # )
                # iter_data_time = time.time()

    return sum(losses) / len(losses)

if __name__ == '__main__':
    print("Start")

    yaml_config = """
            task: segmentation
            class: modelnet.ModelNetDataset
            name: luxnet
            dataroot: resources/light_sim/
            training:
                batch_size: 1
                num_workers: 0
                learning_rate: 0.01
                shuffle: True
                precompute_multi_scale: False
                dataset_size: 80 # in percentage
            testing:
                dataset_size: 20 # in percentage
            # pre_transforms:
            #     - transform: NormalizeScale
            #     - transform: GridSampling3D
            #       lparams: [0.02]
            #       params: 
            #         quantize_coords: [True]                    
            # train_transforms:
            #     - transform: GridSampling3D
            #       lparams: [0.01]
            #       params: 
            #         quantize_coords: [True]
            # test_transforms:
            """

    params = OmegaConf.create(yaml_config)

    GENERATE_DATASET = True
    if GENERATE_DATASET:
        dataset = DeepLuxDataset(params)
        with open(f'data/dataset_v2.pkl', 'wb') as output:
            pickle.dump(dataset, output, pickle.HIGHEST_PROTOCOL)
    else:
        with open(f'data/dataset_v2.pkl', 'rb') as input:
            dataset = pickle.load(input)
    # print(dataset)

    model = Sparseconv3dRegressor()
    # model = RSConvRegressor()

    dataset.create_dataloaders(
        model,
        batch_size=params.training.batch_size,
        shuffle=params.training.shuffle,
        num_workers=params.training.num_workers,
        precompute_multi_scale=params.training.precompute_multi_scale
    )
    print(dataset)

    # print(next(iter(dataset.test_dataloaders[0])))

    # Setup
    if torch.cuda.is_available():
        device = torch.device("cuda:0")
        torch.cuda.set_device(device)
        print("INFO: GPU was set")
    else:
        device = torch.device("cpu")
        print("WARNING: CPU only, this will be slow!")
    optimizer = torch.optim.Adam(model.parameters(), lr=params.training.learning_rate)
    EPOCHS = 20

    # # Setup the tracker and actiavte tensorboard loging
    # logdir = "./runs"
    # logdir = os.path.join(logdir, str(datetime.datetime.now()))
    # os.mkdir(logdir)
    # os.chdir(logdir)
    # tracker = dataset.get_tracker(False, True)

    # initiate writer for tensorboard
    comment = f'batch_size={params.training.batch_size} num_workers={params.training.num_workers} lr={params.training.learning_rate} shuffle={params.training.shuffle} model={"Sparceconv3d"}'
    logdir = "./runs"
    logdir = os.path.join(logdir, str(datetime.datetime.now().strftime('%b%d_%H:%M:%S')), comment)
    modeldir = os.path.join(logdir, 'models')
    writer = SummaryWriter(log_dir=logdir)

    for i in range(EPOCHS):
        print("=========== EPOCH %i ===========" % i)
        time.sleep(0.5)
        # train_epoch(device)
        # tracker.publish(i)
        # # test_epoch('cuda')

        train_loss = train_epoch(device)
        print("Train Loss: {}".format(train_loss))
        torch.cuda.empty_cache()

        test_loss = test_epoch(device)
        print("Test Loss: {}\n".format(test_loss))
        torch.cuda.empty_cache()

        writer.add_scalar("Loss/train", train_loss, i)
        writer.add_scalar("Loss/test", test_loss, i)
        writer.flush()

    print('End')
    os._exit(0)
```

I do not know if you see anything wrong. You could also test it yourself if you want, you can get the dataset files from here https://we.tl/t-DybKFl6O1y  

Bear in mind that my point clouds are of different sizes, unlike your case and my point coordinates are in millimeters.
